Earthfile

```diff
@@ -0,0 +1,126 @@
+VERSION 0.6
+FROM ubuntu:20.04
+
+shared:
+    FROM scratch
+
+    COPY .rubocop.yml .
+    SAVE ARTIFACT .rubocop.yml
+
+    COPY .gitignore .
+    SAVE ARTIFACT .gitignore
+
+deps:
+    ARG DEBIAN_FRONTEND="noninteractive"
+
+    RUN apt-get update \
+     && apt-get upgrade -y \
+     && apt-get install -y --no-install-recommends \
+            build-essential \
+            dirmngr \
+            git \
+            bzr \
+            mercurial \
+            gnupg2 \
+            ca-certificates \
+            curl \
+            file \
+            zlib1g-dev \
+            liblzma-dev \
+            tzdata \
+            zip \
+            unzip \
+            openssh-client \
+            software-properties-common \
+            make \
+     && rm -rf /var/lib/apt/lists/*
+
+docker:
+    FROM +deps
+
+    LABEL org.opencontainers.image.title="dependabot-core"
+    LABEL org.opencontainers.image.source="https://github.com/dependabot/dependabot-core"
+    LABEL org.opencontainers.image.created="$(date -u +'%Y-%m-%dT%H:%M:%SZ')"
+    LABEL org.opencontainers.image.revision="$(git rev-parse HEAD)"
+    LABEL org.opencontainers.image.version="$(git describe --tags --abbrev=0)"
+
+    ARG development
+    ARG tag="latest"
+
+    ENV DEPENDABOT_NATIVE_HELPERS_PATH="${DEPENDABOT_NATIVE_HELPERS_PATH:-/opt}"
+
+    DO ./bundler+SETUP
+    DO ./python+SETUP
+    DO ./npm_and_yarn+SETUP
+
+    # Elm is amd64 only, see:
+    # - https://github.com/elm/compiler/issues/2007
+    # - https://github.com/elm/compiler/issues/2232
+    IF  [ "$TARGETARCH" == "amd64" ]
+        DO ./elm+SETUP
+    END
+
+    DO ./composer+SETUP
+    DO ./go_modules+SETUP
+    DO ./hex+SETUP
+    DO ./cargo+SETUP
+    DO ./terraform+SETUP
+    DO ./pub+SETUP
+
+    DO ./common+CREATE_DEPENDABOT_USER
+
+    COPY --chown=dependabot:dependabot LICENSE /home/dependabot
+
+    WORKDIR /home/dependabot/dependabot-core
+
+    COPY --chown=dependabot:dependabot --dir \
+            omnibus \
+            git_submodules \
+            terraform \
+            github_actions \
+            hex \
+            elm \
+            docker \
+            nuget \
+            maven \
+            gradle \
+            cargo \
+            composer \
+            go_modules \
+            python \
+            pub \
+            npm_and_yarn \
+            bundler \
+            common \
+            .
+
+    USER dependabot
+
+    ENV HOME="/home/dependabot"
+    WORKDIR ${HOME}
+
+    ENTRYPOINT ["/bin/sh"]
+
+    IF [ $development ]
```

Comment:
This isn't about this particular line but I wanted to comment on this line in the PR description:

> No more bin/docker-dev-shell: Rebuilding containers is fast enough that it's no longer necessary to develop within a Docker image.

I'm probably doing something wrong but I'm not finding it fast to rebuild containers. I installed earthly in a 32 core codespace and it seems to do a lot of work each time I run `earthly +docker --development=true`, even if there are  no changes. It's definitely faster than rebuilding from scratch but I'm finding it takes 15s to complete with no changes and a minute if I make a change to the npm readme. Does it run faster for you?

I often have a lot of learning to do when working on core and tend to make a lot of small, iterative changes to debug. If it's going to add more than an extra second or two to rebuild & try a change I think I'd prefer to keep the dev container to keep the feedback loop quick.

I also think this might hamper my current workflow of running `rubocop -A` in the container to fix my poorly styled code? I currently rely on the volume mounting done by the dev shell script to get the edits back into my workspace.

Reply:
> No more bin/docker-dev-shell: Rebuilding containers is fast enough that it's no longer necessary to develop within a Docker image.
> I currently rely on the volume mounting done by the dev shell script to get the edits back into my workspace.

I am behind the curve on reading in on Earthly adoption, apologies - I also wanted to comment on this part of the description.

We have an internal debugging workflow which effectively wraps the ` bin/docker-dev-shell` and volume mounting to allow us to reproduce some issues that involve the production network stack.

It might be possible to just relocate the script into the network debugging tool but it is useful to have it travel with core right now as that means it gets more use/exposure than it would in the debugging kit.

Reply:
@mctofu @brrygrdn Thanks for your feedback about `bin/docker-dev-shell`. I'll admit that I haven't relied on that myself, and don't have a good understanding of how it's being used.

> I'm probably doing something wrong but I'm not finding it fast to rebuild containers. I installed earthly in a 32 core codespace and it seems to do a lot of work each time I run `earthly +docker --development=true`, even if there are no changes. It's definitely faster than rebuilding from scratch but I'm finding it takes 15s to complete with no changes and a minute if I make a change to the npm readme. Does it run faster for you?

On my 2017 iMac, it takes ~20 seconds to rebuild with the `--use-inline-cache` (down from ~1 minute without):

```console
$ time earthly --use-inline-cache +docker --development=true
# ...
________________________________________________________
Executed in   19.51 secs   fish           external 

$ time earthly +docker --development=true
# ...
________________________________________________________
Executed in   56.47 secs   fish           external 
```

How much faster does it run with `--use-inline-cache`?

As a recovering iOS developer, I have a higher tolerance for long compile times. 😅 But I totally get where you're coming from, and want to have a solution that can provide a quick feedback loop.

> I often have a lot of learning to do when working on core and tend to make a lot of small, iterative changes to debug. If it's going to add more than an extra second or two to rebuild & try a change I think I'd prefer to keep the dev container to keep the feedback loop quick.

If we can't get the feedback loop fast enough with cached rebuilds, then we have a couple options:

1. Run the `dependabot-core-development` image with a bind mount, as in `bin/docker-dev-shell`, or
2. Get the same effect developing within Codespaces with `dependabot-core-development` as the base image.

If the `docker run` invocation is cumbersome, we could create a new Earthly target like this:

```earthly
dev-shell:
  LOCALLY

  BUILD (+docker --development=true)

  RUN docker run --rm dependabot-core-development \
        -v $(PWD):/home/dependabot/dependabot-core # etc.
```

> I also think this might hamper my current workflow of running `rubocop -A` in the container to fix my poorly styled code? I currently rely on the volume mounting done by the dev shell script to get the edits back into my workspace.

That's a great call-out. We should be able to support that by doing something like this:

```earthly
rubocop:
    FROM ruby:2.7.1

    RUN gem install rubocop --no-document

    CMD ["rubocop"]
    
    SAVE IMAGE rubocop

lint:
    LOCALLY

    ARG autocorrect
    ARG --required path

    WITH DOCKER --load=+rubocop
        RUN docker run --rm rubocop \
              -v $(PWD):/ \
            -- ${autocorrect:+"-A"} ${path}
    END
```

Then, to run the linter, you'd do:

```console
$ earthly +lint --autocorrect=true --path=bundler
```

Reply:
@mctofu What do you think about @brrygrdn's suggestion to move the dev shell out of core?

> We have an internal debugging workflow which effectively wraps the bin/docker-dev-shell and volume mounting to allow us to reproduce some issues that involve the production network stack.
> 
> It might be possible to just relocate the script into the network debugging tool but it is useful to have it travel with core right now as that means it gets more use/exposure than it would in the debugging kit.

I just updated the `+docker --development=true` target to be closer to our original, but it still copies over sources rather than mounting local ecosystem directories as volumes. Do you think that'll work for your development flow?  

Alternatively, we could continue to ship `Dockerfile.development` and `bin/docker-dev-shell` (since it can still build off a `dependabot-core` image built with Earthly).

Reply:
Unless we can get the time it takes to make a code edit & see it run to within a few seconds I think we should keep the volume mounting support around.

I haven't been a fan of using the core image as a codespace dev container as that gets really cumbersome when you end up needing to change the Dockerfile. I don't think it necessarily needs to be via `Dockerfile.development` and `bin/docker-dev-shell` though. If we do the volume mounts for a dev shell via earthly to get the same result that could be sufficient?

I think it'd also be great if we could rework the `dry-run` script to work more independently on a per ecosystem basis. Currently it supports all ecosystems so if I just want to dry run the go modules updater I still need to build/install a lot of things for the other ecosystems. The dev experience would be much better if getting into a dev shell was quick because it only required building the tooling for that ecosystem.